File: graph_algorithms/week4_fastest_paths/2_detecting_anomalies/negative_cycle.cpp
#include <iostream>
#include <vector>

using std::vector;

int negative_cycle(vector<vector<int>> &adj, vector<vector<int>> &cost) {
  int n = adj.size();
  vector<long long> dist(n, 0);

  // Perform n iterations of relaxation.
  for (int k = 0; k < n; k++) {
    for (int u = 0; u < n; u++) {
      for (int i = 0; i < adj[u].size(); i++) {
        int v = adj[u][i];

        if (dist[v] > dist[u] + cost[u][i])
          dist[v] = dist[u] + cost[u][i];
      }
    }
  }

  // Check for negative cycles.
  for (int u = 0; u < n; u++) {
    for (int i = 0; i < adj[u].size(); i++) {
      int v = adj[u][i];

      if (dist[v] > dist[u] + cost[u][i])
        return 1;
    }
  }

  return 0;
}

int main() {
  int n, m;
  std::cin >> n >> m;

  vector<vector<int>> adj(n, vector<int>());
  vector<vector<int>> cost(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y, w;
    std::cin >> x >> y >> w;
    adj[x - 1].push_back(y - 1);
    cost[x - 1].push_back(w);
  }

  std::cout << negative_cycle(adj, cost) << std::endl;
}

File: graph_algorithms/week4_fastest_paths/2_detecting_anomalies/negative_cycle.py
from typing import List


def negative_cycle(adj: List[List[int]], cost: List[List[int]]) -> int:  # noqa: C901
    n = len(adj)
    dist = [0] * n

    # Perform n iterations of relaxation.
    for _ in range(n):
        for u in range(n):
            for i in range(len(adj[u])):
                v = adj[u][i]

                if dist[v] > dist[u] + cost[u][i]:
                    dist[v] = dist[u] + cost[u][i]

    # Check for negative cycles.
    for u in range(n):
        for i in range(len(adj[u])):
            v = adj[u][i]

            if dist[v] > dist[u] + cost[u][i]:
                return 1

    return 0


if __name__ == "__main__":
    n, m = list(map(int, input().split()))

    edges = [tuple(map(int, input().split())) for _ in range(m)]
    adj: List[List[int]] = [[] for _ in range(n)]
    cost: List[List[int]] = [[] for _ in range(n)]
    for a, b, w in edges:
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)

    print(negative_cycle(adj, cost))
